src/rust/engine/options/src/config.rs

```diff
@@ -1,18 +1,94 @@
 // Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 // Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-use std::collections::HashSet;
+use std::collections::{HashMap, HashSet};
 use std::fs;
 use std::mem;
 use std::path::Path;
 
+use regex::Regex;
 use toml::value::Table;
 use toml::Value;
 
 use super::id::{NameTransform, OptionId};
 use super::parse::parse_string_list;
 use super::{ListEdit, ListEditAction, OptionsSource};
 
+type InterpolationMap = HashMap<String, String>;
+
+pub(crate) fn interpolate_string(
+    value: String,
+    replacements: &InterpolationMap,
+) -> Result<String, String> {
+    let placeholder_re = Regex::new(r"%\(([a-zA-Z0-9_.]+)\)s").unwrap();
```

Comment:
The docs of https://docs.rs/regex/latest/regex/struct.Regex.html#method.new suggest that calling this repeatedly is probably best avoided if possible, as it's not cached. There's a few ways this could happen:

- store it in a `lazy_static::lazy_static! { ... }` so it's created and stored in a global, on first use
- have a context object that stores it, with `interpolate_string` as a value on it, that is used for multiple interpolations

Reply:
Hmm yes, I meant to do this once I realized there was no Python re-style caching, and forgot.

Reply:
Good call. Used a lazy_static (a context object would be annoying here).